wallet/src/main/java/com/android/identity_credential/wallet/ui/destination/provisioncredential/ProvisionCredentialScreen.kt

```diff
@@ -296,4 +343,159 @@ fun ProvisionDocumentScreen(
+}
+
+suspend fun openid4VpPresentation(
+    credential: Credential,
+    walletApp: WalletApplication,
+    fragmentActivity: FragmentActivity,
+    originUri: String,
+    request: String
+): String {
+    val parts = request.split('.')
+    val openid4vpRequest = JSONObject(String(parts[1].fromBase64Url()))
+    val nonceBase64 = openid4vpRequest.getString("nonce")
+    val nonce = Base64.decode(nonceBase64, Base64.NO_WRAP or Base64.URL_SAFE)
+    val clientID = openid4vpRequest.getString("client_id")
+
+    val presentationDefinition = openid4vpRequest.getJSONObject("presentation_definition")
+    val inputDescriptors = presentationDefinition.getJSONArray("input_descriptors")
+    if (inputDescriptors.length() != 1) {
+        throw IllegalArgumentException("Only support a single input input_descriptor")
+    }
+    val inputDescriptor = inputDescriptors.getJSONObject(0)!!
+    val docType = inputDescriptor.getString("id")
+
+    val constraints = inputDescriptor.getJSONObject("constraints")
+    val fields = constraints.getJSONArray("fields")
+
+    val requestedData = mutableMapOf<String, MutableList<Pair<String, Boolean>>>()
+
+    for (n in 0 until fields.length()) {
+        val field = fields.getJSONObject(n)
+        // Only support a single path entry for now
+        val path = field.getJSONArray("path").getString(0)!!
+        // JSONPath is horrible, hacky way to parse it for demonstration purposes
+        val st = StringTokenizer(path, "'", false).asSequence().toList()
+        val namespace = st[1] as String
+        val name = st[3] as String
+        val intentToRetain = field.getBoolean("intent_to_retain")
+        requestedData.getOrPut(namespace) { mutableListOf() }
+            .add(Pair(name, intentToRetain))
+    }
+
+    val consentFields = MdocConsentField.generateConsentFields(
+        docType,
+        requestedData,
+        walletApp.documentTypeRepository,
+        credential as MdocCredential
+    )
+
+    // Generate the Session Transcript
+    val encodedSessionTranscript = CredmanUtil.generateBrowserSessionTranscript(
+        nonce,
+        originUri,
+        Crypto.digest(Algorithm.SHA256, clientID.toByteArray())
+    )
+    val deviceResponse = showPresentmentFlowAndGetDeviceResponse(
+        fragmentActivity,
+        credential,
+        consentFields,
+        null,
+        originUri,
+        encodedSessionTranscript
+    )
+
+    val clientMetadata = openid4vpRequest.getJSONObject("client_metadata")
+    val encryptionAlg = clientMetadata.getString("authorization_encrypted_response_alg")
+    val encryptionEnc = clientMetadata.getString("authorization_encrypted_response_enc")
+
+    // Create the openid4vp response
+    val responseBody = buildJsonObject {
+        if (openid4vpRequest.has("state")) {
+            put("state", openid4vpRequest.getString("state"))
+        }
+        put("vp_token", deviceResponse.toBase64Url())
+
+        // TODO: do we need this?
```

Comment:
Is this still an open question, or can we remove this?

Reply:
Yes, I do not understand if it is required. Our implementation does not need it, but maybe others would. Perhaps once we have a library, it will be solved on the library level.